File: container-with-most-water/container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        
        if(height.empty()) return 0;
        
        int low=0;
        int high=height.size()-1;
        int gmax=INT_MIN;
        int area;
        
        while(low<high){
            area=min(height[high],height[low]) * (high-low);
            gmax=max(gmax,area); 
            if(height[high]>height[low]){
                low++;
            }else{
                high--;
            }
            
                
        }
        return gmax;
        
    }
};

File: 11-container-with-most-water/container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        
        # max_area = 0

        # for l in range(len(height)):
        #     for r in range(l+1,len(height)):
        #         # area = breadth * length
        #         area = (r-l) * min(height[l], height[r])
        #         max_area = max(area, max_area)

        # return max_area
        
        max_area = 0

        left, right = 0, len(height) - 1

        while left < right:
                # area = breadth * length
                area = (right-left) * min(height[left], height[right])
                max_area = max(area, max_area)

                if height[left] < height[right]:
                    left = left + 1
                elif height[left] > height[right]:
                    right = right - 1
                else:
                    # we can move either of the pointer
                    left = left + 1

        return max_area

